### ISSUE TITLE
Failure to propagate non-None type on looped non-None assignment

### ISSUE THREAD

**Issue by jtlapp** [microsoft/pyright#2197]
I don't know how to generically describe this problem. Just look at the code and its error:

```
import random

class Demo:
    def __init__(self):
        self.name = None if random.random() < 0.5 else "str"

    def correct(self):
        if self.name is not None:
            self.name = self.name.replace("a", "b")  # no error here
            for code, correction in {"|": "", "!": ".", "_": " "}.items():
                self.name = self.name.replace(code, correction)
                # ^ ERROR: "replace" is not a known member of "None"

```

This is according to Pylance v2021.8.1 in VSCode.

**Comment by erictraut** [comment]
Thanks for reporting the problem. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.165, which I just published. It will also be included in the next release of pylance.
